src/System.Private.CoreLib/src/System/StubHelpers.cs

```diff
@@ -71,18 +69,17 @@ internal static unsafe IntPtr ConvertToNative(int flags, string strManaged, IntP
                 // If we are marshaling into a stack buffer or we can accurately estimate the size of the required heap
                 // space, we will use a "1-pass" mode where we convert the string directly into the unmanaged buffer.
 
-                // + 1 for the null character from the user
-                nb = (strManaged.Length + 1) * Marshal.SystemMaxDBCSCharSize;
+                // + 1 for the null character from the user.  + 1 for the null character we put in.
+                nb = checked((strManaged.Length + 1) * Marshal.SystemMaxDBCSCharSize + 1);
 
                 // Use the pre-allocated buffer (allocated by localloc IL instruction) if not NULL, 
                 // otherwise fallback to AllocCoTaskMem
                 if (pbNativeBuffer == null)
                 {
-                    // + 1 for the null character we put in
-                    pbNativeBuffer = (byte*)Marshal.AllocCoTaskMem(nb + 1);
+                    pbNativeBuffer = (byte*)Marshal.AllocCoTaskMem(nb);
                 }
 
-                nb = strManaged.ConvertToAnsi(pbNativeBuffer, nb + 1, 0 != (flags & 0xFF), 0 != (flags >> 8));
+                nb = strManaged.ConvertToAnsi(pbNativeBuffer, nb, 0 != (flags & 0xFF), 0 != (flags >> 8));
```

Comment:
Can you doc/comment on the right shift for the flags. What are we checking/validating here?

Reply:
Convention between the IL generated by the VM and these methods. The interface between the IL generated by the VM and the C# helper methods is not designed well. (And there is too much logic in the IL generated by the VM that can be in C# helper methods instead.)

My change is not changing these conventions ... I will add a small comment.